staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -130,32 +157,35 @@ func (q *delayingType) waitingLoop() {
 	// Make a placeholder channel to use when there are no items in our list
 	never := make(<-chan time.Time)
 
+	waitingForQueue := &waitForPriorityQueue{}
+	heap.Init(waitingForQueue)
+
+	waitingEntryByData := map[t]*waitFor{}
+
 	for {
 		if q.Interface.ShuttingDown() {
-			// discard waiting entries
-			q.waitingForAdd = nil
-			q.waitingTimeByEntry = nil
 			return
 		}
 
 		now := q.clock.Now()
 
 		// Add ready entries
-		readyEntries := 0
-		for _, entry := range q.waitingForAdd {
+		for waitingForQueue.Len() > 0 {
+			entry := waitingForQueue.Peek().(*waitFor)
```

Comment:
@deads2k Sounds good. I've added a commit with some documentation. How do you think it reads?